File: src/api/file.ts
import { invoke } from '@tauri-apps/api/core'
import { open, save } from '@tauri-apps/plugin-dialog'

/**
 * File API - Frontend interface to file system operations
 */
export const fileAPI = {
  /**
   * Show file open dialog
   */
  async showOpenDialog(): Promise<string | null> {
    const selected = await open({
      multiple: false,
      directory: false,
      filters: [{
        name: 'EDA Project',
        extensions: ['eda', 'json']
      }]
    })
    return selected as string | null
  },

  /**
   * Show file save dialog
   */
  async showSaveDialog(defaultName?: string): Promise<string | null> {
    const selected = await save({
      defaultPath: defaultName,
      filters: [{
        name: 'EDA Project',
        extensions: ['eda', 'json']
      }]
    })
    return selected
  },

  /**
   * Read file contents as string
   */
  async readFile(path: string): Promise<string> {
    return await invoke<string>('read_file', { path })
  },

  /**
   * Write string content to file
   */
  async writeFile(path: string, content: string): Promise<void> {
    return await invoke<void>('write_file', { path, content })
  },

  /**
   * Check if file exists
   */
  async fileExists(path: string): Promise<boolean> {
    return await invoke<boolean>('file_exists', { path })
  },

  /**
   * Create directory (including parent directories)
   */
  async createDirectory(path: string): Promise<void> {
    return await invoke<void>('create_directory', { path })
  },

  /**
   * Get file information (size, modified date, etc.)
   */
  async getFileInfo(path: string): Promise<{
    size: number
    modified: string
    isDirectory: boolean
  }> {
    return await invoke('get_file_info', { path })
  }
}

File: src-tauri/src/commands/file.rs
use std::path::PathBuf;

#[tauri::command]
pub async fn show_open_dialog() -> Result<Option<PathBuf>, String> {
    // This will be handled by Tauri's dialog API on the frontend
    Ok(None)
}

#[tauri::command]
pub async fn show_save_dialog() -> Result<Option<PathBuf>, String> {
    // This will be handled by Tauri's dialog API on the frontend
    Ok(None)
}

#[tauri::command]
pub async fn read_file(path: String) -> Result<String, String> {
    if !crate::utils::validators::validate_file_path(&path) {
        return Err("Invalid file path".to_string());
    }
    
    std::fs::read_to_string(&path)
        .map_err(|e| format!("Failed to read file: {}", e))
}

#[tauri::command]
pub async fn write_file(path: String, content: String) -> Result<(), String> {
    if !crate::utils::validators::validate_file_path(&path) {
        return Err("Invalid file path".to_string());
    }
    
    std::fs::write(&path, content)
        .map_err(|e| format!("Failed to write file: {}", e))
}

#[tauri::command]
pub async fn file_exists(path: String) -> Result<bool, String> {
    Ok(std::path::Path::new(&path).exists())
}

#[tauri::command]
pub async fn create_directory(path: String) -> Result<(), String> {
    if !crate::utils::validators::validate_file_path(&path) {
        return Err("Invalid directory path".to_string());
    }
    
    std::fs::create_dir_all(&path)
        .map_err(|e| format!("Failed to create directory: {}", e))
}

#[tauri::command]
pub async fn get_file_info(path: String) -> Result<FileInfo, String> {
    let metadata = std::fs::metadata(&path)
        .map_err(|e| format!("Failed to get file info: {}", e))?;
    
    Ok(FileInfo {
        size: metadata.len(),
        is_directory: metadata.is_dir(),
        is_file: metadata.is_file(),
        modified: metadata.modified()
            .ok()
            .and_then(|t| t.duration_since(std::time::UNIX_EPOCH).ok())
            .map(|d| d.as_secs()),
    })
}

#[derive(serde::Serialize)]
#[serde(rename_all = "camelCase")]
pub struct FileInfo {
    pub size: u64,
    pub is_directory: bool,
    pub is_file: bool,
    pub modified: Option<u64>,
}